extensions/android-paging3/src/main/java/com/squareup/sqldelight/android/paging3/OffsetQueryPagingSource.kt

```diff
@@ -34,9 +34,13 @@ internal class OffsetQueryPagingSource<RowType : Any>(
     params: LoadParams<Long>
   ): LoadResult<Long, RowType> = withContext(dispatcher) {
     try {
-      val key = params.key ?: 0L
       transacter.transactionWithResult {
         val count = countQuery.executeAsOne()
+        val key = when (params) {
+          is LoadParams.Refresh -> params.key?.let { notNullKey -> minOf(count - 1, notNullKey) }
+          else -> params.key
+        } ?: 0L
```

Comment:
For Refresh you may additionally want to offset the initial key to be at most `count - params.loadSize`. This will let you load the most items in the initial load, without needing to trigger prepend to fulfill PagingConfig.initialLoadSize. It will also increase the chance DiffUtil will find a matching anchorPosition when it receives the initial load result after invalidation.

Reply:
Awesome! Thank you for the input :)

Reply:
Unfortunately that coerceIn will always allocate the IntRange. The two-argument variant will not, although you'll still have to do the `- 1` I believe.

Reply:
I don't know why I just assumed ranges are value classes / inlinable. Jake is right to avoid the allocation!